File: source_code/matrix.py
import os
class SparseMatrix:
    def __init__(self, file_path=None, num_rows=0, num_cols=0):
        # initialize a sparse matrix
        self.rows = 0
        self.cols = 0
        self.data = {} #to store non zero elements
        if file_path is not None:
            self._load_from_file(file_path)
        else:
            self.rows = num_rows
            self.cols = num_cols

    def _load_from_file(self, file_path):
        try:
            with open(file_path, 'r') as f:
                lines = [line.strip() for line in f if line.strip()]
                if not lines[0].startswith('rows=') or not lines[1].startswith('cols='):
                    raise ValueError("Input file has wrong format")
                self.rows = int(lines[0][5:])
                self.cols = int(lines[1][5:])
                for line in lines[2:]:
                    if not (line.startswith('(') and line.endswith(')')):
                        raise ValueError("Input file has wrong format")
                    p = line[1:-1].split(',')
                    if len(p) != 3:
                        raise ValueError("Input file has wrong format")
                    try:
                        r = int(p[0].strip())
                        c = int(p[1].strip())
                        v = int(p[2].strip())
                    except ValueError:
                        raise ValueError("Input file has wrong format")
                    try:
                        self.set_element(r, c, v)
                    except IndexError:
                        continue
        except Exception as e:
            raise ValueError(f"Error reading file: {str(e)}")

    def get_element(self, r, c):
        if r >= self.rows or c >= self.cols or r < 0 or c < 0:
            raise IndexError("Index out of bounds")
        return self.data.get((r, c), 0)

    def set_element(self, r, c, v):
        if r >= self.rows or c >= self.cols or r < 0 or c < 0:
            raise IndexError("Index out of bounds")
        if v == 0:
            self.data.pop((r, c), None)
        else:
            self.data[(r, c)] = v

    def add(self, other):
        if self.rows != other.rows or self.cols != other.cols:
            raise ValueError("Matrix dimensions must match for addition")
        res = SparseMatrix(num_rows=self.rows, num_cols=self.cols)
        for (r, c), v in self.data.items():
            res.set_element(r, c, v)
        for (r, c), v in other.data.items():
            cur = res.get_element(r, c)
            res.set_element(r, c, cur + v)
        return res

    def subtract(self, other):
        if self.rows != other.rows or self.cols != other.cols:
            raise ValueError("Matrix dimensions must match for subtraction")
        res = SparseMatrix(num_rows=self.rows, num_cols=self.cols)
        for (r, c), v in self.data.items():
            res.set_element(r, c, v)
        for (r, c), v in other.data.items():
            cur = res.get_element(r, c)
            res.set_element(r, c, cur - v)
        return res

    def multiply(self, other):
        if self.cols != other.rows:
            raise ValueError("Number of columns in first matrix must match number of rows in second matrix")
        res = SparseMatrix(num_rows=self.rows, num_cols=other.cols)
        other_map = {}
        for (r, c), v in other.data.items():
            if r not in other_map:
                other_map[r] = []
            other_map[r].append((c, v))
        for (i, k), a in self.data.items():
            if k in other_map:
                for (j, b) in other_map[k]:
                    cur = res.get_element(i, j)
                    res.set_element(i, j, cur + a * b)
        return res

    def save_to_file(self, file_path):
        with open(file_path, 'w') as f:
            f.write(f"rows={self.rows}\n")
            f.write(f"cols={self.cols}\n")
            for (r, c), v in sorted(self.data.items()):
                f.write(f"({r}, {c}, {v})\n")

def main():
    # function to handle user input and perform operations
    os.makedirs("results", exist_ok=True) #creates results directory if it doen't exist
    print("Sparse Matrix Operations(add, subtract, multiply)")
    print("1. Add")
    print("2. Subtract")
    print("3. Multiply")
    try:
        ch = int(input("Enter your choice (1-3): "))
        if ch not in [1, 2, 3]:
            raise ValueError("Invalid choice")
        f1 = input("Enter path to first matrix file: ")
        f2 = input("Enter path to second matrix file: ")
        m1 = SparseMatrix(f1)
        m2 = SparseMatrix(f2)
        if ch == 1:
            res = m1.add(m2)
            out_file = "results/addition_result.txt"
        elif ch == 2:
            res = m1.subtract(m2)
            out_file = "results/subtraction_result.txt"
        elif ch == 3:
            res = m1.multiply(m2)
            out_file = "results/multiplication_result.txt"
        res.save_to_file(out_file)
        print(f"Operation completed. Results saved to: {os.path.abspath(out_file)}")
    except Exception as e:
        print(f"Error: {str(e)}")

if __name__ == "__main__":
    main()

File: source_code/matrix.js
import fs from "fs";

class SparseMatrix {
  constructor(numRows, numCols) {
    this.rows = numRows;
    this.cols = numCols;
    this.data = new Map();
  }

  static async fromFile(filePath) {
    try {
      const content = await fs.promises.readFile(filePath, "utf8");

      const lines = content.split("\n");

      const rows = parseInt(lines[0].substring(5));

      const cols = parseInt(lines[1].substring(5));

      const matrix = new SparseMatrix(rows, cols);

      for (let i = 2; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;

        const values = line
          .slice(1, -1)
          .split(",")
          .map((x) => parseInt(x.trim()));
        matrix.setElement(values[0], values[1], values[2]);
      }

      return matrix;
    } catch (error) {
      throw new Error(`Failed to read matrix file: ${filePath}\nError details: ${error.message}`);
    }
  }

  getElement(row, col) {
    const rowData = this.data.get(row);
    return rowData ? rowData.get(col) || 0 : 0;
  }

  setElement(row, col, value) {
    if (value === 0) {
      const rowData = this.data.get(row);

      if (rowData) {
        rowData.delete(col);
        if (rowData.size === 0) {
          this.data.delete(row);
        }
      }
    } else {
      if (!this.data.has(row)) {
        this.data.set(row, new Map());
      }
      this.data.get(row).set(col, value);
    }
  }

  getDimensionString() {
    return `${this.rows}x${this.cols}`;
  }

  add(other) {
    if (this.rows !== other.rows || this.cols !== other.cols) {
      throw new Error(`Matrix dimensions must match for addition. Matrix 1 is ${this.rows}x${this.cols}, Matrix 2 is ${other.rows}x${other.cols}`);
    }

    const result = new SparseMatrix(this.rows, this.cols);

    for (const [row, rowData] of this.data) {
      for (const [col, value] of rowData) {
        result.setElement(row, col, value);
      }
    }
    for (const [row, rowData] of other.data) {
      for (const [col, value] of rowData) {
        const sum = result.getElement(row, col) + value;
        result.setElement(row, col, sum);
      }
    }

    return result;
  }

  subtract(other) {
    if (this.rows !== other.rows || this.cols !== other.cols) {
      throw new Error(`Matrix dimensions must match for subtraction. Matrix 1 is ${this.rows}x${this.cols}, Matrix 2 is ${other.rows}x${other.cols}`);
    }

    const result = new SparseMatrix(this.rows, this.cols);
    for (const [row, rowData] of this.data) {
      for (const [col, value] of rowData) {
        result.setElement(row, col, value);
      }
    }

    for (const [row, rowData] of other.data) {
      for (const [col, value] of rowData) {
        const diff = result.getElement(row, col) - value;
        result.setElement(row, col, diff);
      }
    }

    return result;
  }

  multiply(other) {
    if (this.cols !== other.rows) {
      throw new Error(`Invalid dimensions for multiplication. Matrix 1 is ${this.rows}x${this.cols}, Matrix 2 is ${other.rows}x${other.cols}. The number of columns in first matrix (${this.cols}) must equal the number of rows in second matrix (${other.rows})`);
    }

    const result = new SparseMatrix(this.rows, other.cols);

    // Only iterate over non-zero elements
    for (const [row1, rowData1] of this.data) {
      for (const [col1, val1] of rowData1) {
        const rowData2 = other.data.get(col1);
        if (rowData2) {
          for (const [col2, val2] of rowData2) {
            const current = result.getElement(row1, col2);
            result.setElement(row1, col2, current + val1 * val2);
          }
        }
      }
    }

    return result;
  }
  toString() {
    let result = `rows=${this.rows}\ncols=${this.cols}\n`;
    for (const [row, rowData] of this.data) {
      for (const [col, value] of rowData) {
        result += `(${row}, ${col}, ${value})\n`;
      }
    }
    return result;
  }
}

async function main() {
  if (process.argv.length < 6) {    console.log(
      "Usage: node main.js <operation> <matrix1_file> <matrix2_file> <output_file>\n" +
      "Operations available: add, subtract, multiply\n" +
      "Example: node main.js add matrix1.txt matrix2.txt result.txt"
    );
    process.exit(1);
  }

  try {
    const operation = process.argv[2];
    const matrix1 = await SparseMatrix.fromFile(process.argv[3]);
    const matrix2 = await SparseMatrix.fromFile(process.argv[4]);
    const outputFile = process.argv[5];    let result;
    // Display the operation with matrix dimensions
    const dim1 = matrix1.getDimensionString();
    const dim2 = matrix2.getDimensionString();
    console.log(`Performing operation: ${dim1} ${operation} ${dim2}`);

    switch (operation) {
      case "add":
        result = matrix1.add(matrix2);
        break;
      case "subtract":
        result = matrix1.subtract(matrix2);
        break;
      case "multiply":
        result = matrix1.multiply(matrix2);
        break;
      default:
        throw new Error(`Invalid operation "${operation}". Valid operations are: add, subtract, or multiply`);
    }

    // Write result to output file
    await fs.promises.writeFile(outputFile, result.toString());
    console.log(`Operation "${operation}" completed successfully. Result written to: ${outputFile}`);
  } catch (error) {
    console.error("Error:", error.message);
    process.exit(1);
  }
}

await main();
